demos/fleet_provisioning/fleet_provisioning_with_csr/pkcs11_operations.c

```diff
@@ -0,0 +1,1162 @@
+        status = ( ret == CKR_OK );
+    }
+
+    if( status == true )
+    {
+        ret = provisionCertificate( p11Session, ( uint8_t * ) claimCert,
+                                    claimCertLength + 1,
+                                    ( uint8_t * ) pkcs11configLABEL_CLAIM_CERTIFICATE );
+        status = ( ret == CKR_OK );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static int extractEcPublicKey( CK_SESSION_HANDLE p11Session,
+                               mbedtls_ecdsa_context * pEcdsaContext,
+                               CK_OBJECT_HANDLE publicKey )
+{
+    CK_ATTRIBUTE ecTemplate = { 0 };
+    int mbedtlsRet = -1;
+    CK_RV pkcs11ret = CKR_OK;
+    CK_BYTE ecPoint[ 67 ] = { 0 };
+    CK_FUNCTION_LIST_PTR pP11FunctionList;
+
+    mbedtls_ecdsa_init( pEcdsaContext );
+    mbedtls_ecp_group_init( &( pEcdsaContext->grp ) );
+
+    pkcs11ret = C_GetFunctionList( &pP11FunctionList );
+
+    if( pkcs11ret != CKR_OK )
+    {
+        LogError( ( "Failed to extract EC public key. Could not get a "
+                    "PKCS #11 function pointer." ) );
+    }
+    else
+    {
+        ecTemplate.type = CKA_EC_POINT;
+        ecTemplate.pValue = ecPoint;
+        ecTemplate.ulValueLen = sizeof( ecPoint );
+        pkcs11ret = pP11FunctionList->C_GetAttributeValue( p11Session, publicKey, &ecTemplate, 1 );
+
+        if( pkcs11ret != CKR_OK )
+        {
+            LogError( ( "Failed to extract EC public key. Could not get attribute value. "
+                        "C_GetAttributeValue failed with %lu.", pkcs11ret ) );
+        }
+    }
+
+    if( pkcs11ret == CKR_OK )
+    {
+        mbedtlsRet = mbedtls_ecp_group_load( &( pEcdsaContext->grp ), MBEDTLS_ECP_DP_SECP256R1 );
+
+        if( mbedtlsRet != 0 )
+        {
+            LogError( ( "Failed creating an EC key. "
+                        "mbedtls_ecp_group_load failed: MbedTLS"
+                        "error = %s : %s.",
+                        mbedtlsHighLevelCodeOrDefault( mbedtlsRet ),
+                        mbedtlsLowLevelCodeOrDefault( mbedtlsRet ) ) );
+            pkcs11ret = CKR_FUNCTION_FAILED;
+        }
+        else
+        {
+            mbedtlsRet = mbedtls_ecp_point_read_binary( &( pEcdsaContext->grp ), &( pEcdsaContext->Q ), &ecPoint[ 2 ], ecTemplate.ulValueLen - 2 );
+
+            if( mbedtlsRet != 0 )
+            {
+                LogError( ( "Failed creating an EC key. "
+                            "mbedtls_ecp_group_load failed: MbedTLS"
+                            "error = %s : %s.",
+                            mbedtlsHighLevelCodeOrDefault( mbedtlsRet ),
+                            mbedtlsLowLevelCodeOrDefault( mbedtlsRet ) ) );
+                pkcs11ret = CKR_FUNCTION_FAILED;
+            }
+        }
+    }
+
+    return mbedtlsRet;
+}
+
+/*-----------------------------------------------------------*/
+
+static int32_t privateKeySigningCallback( void * pContext,
```

Comment:
This function can also be commonized between the transport library and this file

Reply:
This can be done later.